engine/src/main/java/io/camunda/zeebe/engine/state/migration/DbMigrationState.java

```diff
@@ -0,0 +1,85 @@
+import io.camunda.zeebe.engine.state.mutable.MutableTransientProcessMessageSubscriptionState;
+
+public class DbMigrationState implements MutableMigrationState {
+
+  // ZbColumnFamilies.PROCESS_SUBSCRIPTION_BY_SENT_TIME,
+  // (sentTime, elementInstanceKey, messageName) => \0
+  private final DbLong processSubscriptionSentTime;
+  private final DbLong processSubscriptionElementInstanceKey;
+  private final DbString processSubscriptionMessageName;
+
+  private final DbCompositeKey<DbLong, DbString> processSubscriptionElementKeyAndMessageName;
+  private final DbCompositeKey<DbLong, DbCompositeKey<DbLong, DbString>>
+      processSubscriptionSentTimeCompositeKey;
+  private final ColumnFamily<DbCompositeKey<DbLong, DbCompositeKey<DbLong, DbString>>, DbNil>
+      processSubscriptionSentTimeColumnFamily;
+
+  public DbMigrationState(
+      final ZeebeDb<ZbColumnFamilies> zeebeDb, final TransactionContext transactionContext) {
+
+    processSubscriptionElementInstanceKey = new DbLong();
+    processSubscriptionMessageName = new DbString();
+    processSubscriptionElementKeyAndMessageName =
+        new DbCompositeKey<>(processSubscriptionElementInstanceKey, processSubscriptionMessageName);
+
+    processSubscriptionSentTime = new DbLong();
+    processSubscriptionSentTimeCompositeKey =
+        new DbCompositeKey<>(
+            processSubscriptionSentTime, processSubscriptionElementKeyAndMessageName);
+    processSubscriptionSentTimeColumnFamily =
+        zeebeDb.createColumnFamily(
+            ZbColumnFamilies.PROCESS_SUBSCRIPTION_BY_SENT_TIME,
+            transactionContext,
+            processSubscriptionSentTimeCompositeKey,
+            DbNil.INSTANCE);
+  }
+
+  @Override
+  public void migrateProcessMessageSubscriptionSentTime(
+      final MutableProcessMessageSubscriptionState persistentSate,
+      final MutableTransientProcessMessageSubscriptionState transientState) {
+
+    processSubscriptionSentTimeColumnFamily.forEach(
+        (key, value) -> {
+          final var sentTime = key.getFirst().getValue();
+          final var elementKeyAndMessageName = key.getSecond();
+          final var elementInstanceKey = elementKeyAndMessageName.getFirst().getValue();
+          final var messageName = elementKeyAndMessageName.getSecond().getBuffer();
+
+          final var processMessageSubscription =
+              persistentSate.getSubscription(elementInstanceKey, messageName);
+          if (processMessageSubscription != null) {
+
+            final var record = processMessageSubscription.getRecord();
+            if (processMessageSubscription.isOpening()) {
+              persistentSate.put(elementInstanceKey, record);
+              transientState.updateSentTime(record, sentTime);
+            } else if (processMessageSubscription.isClosing()) {
+              persistentSate.updateToClosingState(record);
+              transientState.updateSentTime(record, sentTime);
+            }
```

Comment:
It is technically necessary, because there is no `add()` method defined for the transient state. So I call the methods on the persistent state because I know that as implementation detail they will modify the transient state. 

Now that I put it like this, I find it strange as well. However, I would like to keep the transient interface narrow. 

Personally, I would rather keep it in this strange way, maybe add a comment. Rather than opening the interface wider for a use case that is only relevant during migration. My hope is also that we can retire all the migration code once we no longer support the release. So all migration code for me is only a guest in the code base.

What do you think?

Reply:
I see your point. Let's just add a comment.